### PR TITLE
Implement pyfunc improvements

### PR DESCRIPTION
**Pull request by harupy**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Makes the following improvements:

- Support logging a function as a model
- Infer a model signature from type annotations

## How is this patch tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/types/utils.py [hunk 1]
```diff
@@ -286,3 +289,16 @@ def _is_spark_df(x) -> bool:
...
     except ImportError:
         return False
+
+
+def _infer_schema_from_type_hint(type_hint, example=None):
+    if type_hint == List[str]:
+        return Schema([ColSpec(type="string", name=None)])
+    elif type_hint == List[Dict[str, str]]:
+        if example:
+            return Schema([ColSpec(type="string", name=name) for name in example[0]])
```
Comment (BenWilson2) [inline mlflow/types/utils.py:299]: Do we want the name to be the key of the dict?

Comment (BenWilson2) [reply mlflow/types/utils.py:299]: Also, if the example contains more than 1 dictionary within the list, should we validate that all keys match among all dictionaries?

File: tests/pyfunc/test_model_export_with_class_and_artifacts.py [hunk 2]
```diff
@@ -1168,3 +1171,152 @@ def test_dependency_inference_does_not_exclude_mlflow_dependencies(tmp_path):
...
+        {"name": "d", "type": "string"},
+    ]
+
+
+def test_function_python_model_list_dict_to_list_dict_with_example_pep585(tmp_path):
+    def python_model(x: list[dict[str, str]]) -> list[dict[str, str]]:
+        return [{v: k for k, v in d.items()} for d in x]
+
+    mlflow.pyfunc.save_model(
+        path=tmp_path, python_model=python_model, input_example=[{"a": "b", "c": "d"}]
```
Comment (BenWilson2) [inline tests/pyfunc/test_model_export_with_class_and_artifacts.py:1282]: Could we add a test that processes an example with mixed keys in subsequent dicts within the list? (Just to make sure that we're throwing if a user inputs invalid ColSpec schemas?) 
i.e., `input_example=[{"a": "b"}, {"b": "b"}]`

File: mlflow/types/utils.py (hunk 1 above)
Comment (harupy) [reply mlflow/types/utils.py:299]: > Do we want the name to be the key of the dict?

`name` represents the dict key :)

> Also, if the example contains more than 1 dictionary within the list, should we validate that all keys match among all dictionaries?

ok, will add that validation

File: mlflow/types/utils.py [hunk 3]
```diff
@@ -286,3 +290,107 @@ def _is_spark_df(x) -> bool:
...
+        _validate_is_dict(example)
+        _validate_has_items(example)
+        if num_items is not None:
+            _validate_num_items(example, num_items)
+        _validate_all_keys_string(example)
+        _validate_all_values_string(example)
+        _validate_keys_match(example, first_keys)
+
+
+def _infer_schema_from_type_hint(type_hint, examples=None):
```
Comment (BenWilson2) [inline mlflow/types/utils.py:380]: This is great :)

**Review by BenWilson2 — APPROVED** [review]
One tiny question :) otherwise LGTM!

File: mlflow/pyfunc/model.py [hunk 4]
```diff
@@ -96,6 +96,47 @@ def predict(self, context, model_input):
...
+
+    def __init__(self, func, hints=None, signature=None):
+        self.func = func
+        self.hints = hints
+        self.signature = signature
+
+    def _convert_input(self, model_input):
+        import pandas as pd
+
+        if self.hints.input == List[str]:
```
Comment (harupy) [inline mlflow/pyfunc/model.py:113]: Expected scenario:

```python
def predict(input: List[str]) -> List[str]:
    ...

mlflow.pyfunc.log_model(python_model=predict)

pdf = pd.DataFrame({"str": ["a", "b", "c"]})
model = mlflow.pyfunc.load_model(...)
mode.predict(pdf)
```

File: mlflow/pyfunc/model.py [hunk 5]
```diff
@@ -96,6 +96,47 @@ def predict(self, context, model_input):
...
+            if isinstance(model_input, pd.DataFrame):
+                first_string_column = model_input.select_dtypes(include=["object"]).columns[0]
+                return model_input[first_string_column].tolist()
```
Comment (harupy) [inline mlflow/pyfunc/model.py:116]: Throw a more helpful error instead of `IndexError`.

File: mlflow/models/signature.py [hunk 6]
```diff
@@ -130,3 +133,82 @@ def infer_signature(
...
+    def __init__(self, input_=None, output=None):
+        self.input = input_
+        self.output = output
+
+
+def _extract_type_hints(f):
+    if not hasattr(f, "__annotations__") and hasattr(f, "__call__"):
+        return _extract_type_hints(f.__call__)
+
+    if f.__annotations__ == {}:
```
Comment (BenWilson2) [inline mlflow/models/signature.py:178]: Do we need to add type hints to `mlflow.models.Model` to get this to work?

Comment (BenWilson2) [reply mlflow/models/signature.py:178]: Specifically, prior to py3.10, don't we have to access annotations on a class by doing something like this:
```
if isinstance(f, type):
  annotations = f.__dict__.get("__annotations__", None)
else:
  annotations = getattr(f, "__annotations__", None)
  ```
  
  There's some logic on this in this module note: https://docs.python.org/3/howto/annotations.html

**Review by jinzhang21 — APPROVED** [review]
Awesome PR, @harupy !

File: mlflow/models/signature.py (hunk 6 above)
Comment (harupy) [reply mlflow/models/signature.py:178]: Do we extract type hints from `mlflow.models.Model`?

File: tests/pyfunc/test_model_export_with_class_and_artifacts.py [hunk 7]
```diff
@@ -1168,3 +1175,238 @@ def test_dependency_inference_does_not_exclude_mlflow_dependencies(tmp_path):
...
+    assert model.signature.outputs.to_dict() == [{"type": "string"}]
+
+
+def test_function_python_model_list_dict_to_list_without_example(tmp_path):
+    def python_model(x: List[Dict[str, str]]) -> List[str]:
+        return [str(d) for d in x]
+
+    mlflow.pyfunc.save_model(path=tmp_path, python_model=python_model)
+    model = Model.load(tmp_path)
+    assert model.signature.inputs.to_dict() == [{"type": "string"}]
```
[the remark this answers is not in the document]
Comment (harupy) [reply tests/pyfunc/test_model_export_with_class_and_artifacts.py:1246]: I think we can use an unnamed string `ColSpec`.

**Pull request merged**